channeldb/payment_control.go

```diff
@@ -184,17 +205,57 @@ func (p *PaymentControl) RegisterAttempt(paymentHash lntypes.Hash,
 			return err
 		}
 
-		// Ensure the payment is in-flight and passes the checks set
-		// above.
+		// Ensure the payment is in-flight.
 		if err := ensureInFlight(payment); err != nil {
 			return err
 		}
 
+		// We cannot register a new attempt if the payment already has
+		// reached a terminal condition:
 		settle, fail := payment.TerminalInfo()
 		if settle != nil || fail != nil {
 			return ErrPaymentTerminal
 		}
 
+		// Make sure any existing shards match the new one with regards
+		// to MPP options.
+		mpp := attempt.Route.FinalHop().MPP
+		for _, h := range payment.InFlightHTLCs() {
+			hMpp := h.Route.FinalHop().MPP
+
+			switch {
+
+			// We tried to register a non-MPP attempt for a MPP
+			// payment.
+			case mpp == nil && hMpp != nil:
+				return ErrMPPayment
+
+			// We tried to register a MPP shard for a non-MPP
+			// payment.
+			case mpp != nil && hMpp == nil:
+				return ErrNonMPPayment
+
+			// Non-MPP payment, nothing more to validate.
+			case mpp == nil:
+				continue
```

Comment:
There is a non-mpp inflight and we allow sending another non-mpp? Or is that caught below in the amt check?

Reply:
Added exact amount check